src/types/mod.rs

```diff
@@ -424,6 +424,76 @@ impl ValueInternal for JsNumber {
     }
 }
 
+/// A JavaScript Date object.
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+#[cfg(feature = "napi-runtime")]
+pub struct JsDate(raw::Local);
+
+#[cfg(feature = "napi-runtime")]
+impl Value for JsDate { }
+
+#[cfg(feature = "napi-runtime")]
+impl Managed for JsDate {
+    fn to_raw(self) -> raw::Local { self.0 }
+
+    fn from_raw(_: Env, h: raw::Local) -> Self { JsDate(h) }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum DateErrorKind {
+    Overflow,
+    Underflow,
+}
+
+#[cfg(feature = "napi-runtime")]
+impl JsDate {
+    pub const MIN_VALUE: f64 = -8.64e15;
+    pub const MAX_VALUE: f64 = 8.64e15;
+
+    pub fn new<'a, C: Context<'a>, T: Into<f64>>(cx: &mut C, time: T) -> Handle<'a, JsDate> {
+        let env = cx.env().to_raw();
+        let local = unsafe {
+            neon_runtime::date::new_date(env, time.into())
+        };
+        Handle::new_internal(JsDate(local))
+    }
+
+    pub fn try_new<'a, C: Context<'a>, V: Into<f64> + std::cmp::PartialOrd>(cx: &mut C, value: V) -> Result<Handle<'a, JsDate>, DateErrorKind> {
+        let env = cx.env().to_raw();
+        let local = unsafe {
+            neon_runtime::date::new_date(env, value.into())
+        };
+        let date = Handle::new_internal(JsDate(local));
+        if date.is_valid(cx) { return Ok(date); }
+        if date.value(cx) > JsDate::MAX_VALUE { Err(DateErrorKind::Overflow) } else { Err(DateErrorKind::Underflow) }
```

Comment:
This check should occur as the first step if this function. We aren't guaranteed that `date.value(cx)` will return a non-NaN value.

Separately, this function will return `Ok` if the user passes `NaN`. That seems correct to me, but I wanted to confirm others agree.

Reply:
Good catch! What is the expected behavior when NaN is passed to construct a date?

Reply:
In my opinion, the expected behavior is to successfully return a JS `Date(Invalid Date)` since it wouldn't result in data loss (Rust doesn't differentiate different NaN values).